servers/src/main/java/tachyon/worker/block/TieredBlockStore.java

```diff
@@ -0,0 +1,432 @@
+
+  /** A readwrite lock for meta data **/
+  private final ReentrantReadWriteLock mEvictionLock = new ReentrantReadWriteLock();
+
+  public TieredBlockStore(TachyonConf tachyonConf) throws IOException {
+    mTachyonConf = Preconditions.checkNotNull(tachyonConf);
+    mMetaManager = BlockMetadataManager.newBlockMetadataManager(mTachyonConf);
+    mLockManager = new BlockLockManager(mMetaManager);
+
+    AllocatorType allocatorType =
+        mTachyonConf.getEnum(Constants.WORKER_ALLOCATE_STRATEGY_TYPE, AllocatorType.DEFAULT);
+    mAllocator = AllocatorFactory.create(allocatorType, mMetaManager);
+
+    EvictorType evictorType =
+        mTachyonConf.getEnum(Constants.WORKER_EVICT_STRATEGY_TYPE, EvictorType.DEFAULT);
+    mEvictor = EvictorFactory.create(evictorType, mMetaManager);
+    if (mEvictor instanceof BlockStoreEventListener) {
+      registerBlockStoreEventListener((BlockStoreEventListener) mEvictor);
+    }
+  }
+
+  @Override
+  public long lockBlock(long userId, long blockId) throws IOException {
+    return mLockManager.lockBlock(userId, blockId, BlockLockType.READ);
+  }
+
+  @Override
+  public void unlockBlock(long lockId) throws IOException {
+    mLockManager.unlockBlock(lockId);
+  }
+
+  @Override
+  public void unlockBlock(long userId, long blockId) throws IOException {
+    mLockManager.unlockBlock(userId, blockId);
+  }
+
+  @Override
+  public BlockWriter getBlockWriter(long userId, long blockId) throws IOException {
+    TempBlockMeta tempBlockMeta = mMetaManager.getTempBlockMeta(blockId);
+    return new LocalFileBlockWriter(tempBlockMeta);
+  }
+
+  @Override
+  public BlockReader getBlockReader(long userId, long blockId, long lockId) throws IOException {
+    mLockManager.validateLockId(userId, blockId, lockId);
+    BlockMeta blockMeta = mMetaManager.getBlockMeta(blockId);
+    return new LocalFileBlockReader(blockMeta);
+  }
+
+  @Override
+  public TempBlockMeta createBlockMeta(long userId, long blockId, BlockStoreLocation location,
+      long initialBlockSize) throws IOException {
+    TempBlockMeta tempBlock;
+    mEvictionLock.readLock().lock();
+    try {
+      tempBlock = createBlockMetaNoLock(userId, blockId, location, initialBlockSize);
+    } finally {
+      mEvictionLock.readLock().unlock();
+    }
+    return tempBlock;
+  }
+
+  @Override
+  public BlockMeta getBlockMeta(long userId, long blockId, long lockId) throws IOException {
+    mLockManager.validateLockId(userId, blockId, lockId);
+    return mMetaManager.getBlockMeta(blockId);
+  }
+
+  @Override
+  public void commitBlock(long userId, long blockId) throws IOException {
+    mEvictionLock.readLock().lock();
+    TempBlockMeta tempBlockMeta = mMetaManager.getTempBlockMeta(blockId);
+    try {
+      commitBlockNoLock(userId, blockId, tempBlockMeta);
+      for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
+        listener.onCommitBlock(userId, blockId, tempBlockMeta.getBlockLocation());
+      }
```

Comment:
This possibly can go after the readlock is released, maybe add a TODO to evaluate this?